cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -68,19 +85,54 @@ public void run(FullHttpRequest request, HttpResponder responder) {
     }
 
     String className = null;
+    CountDownLatch latch = new CountDownLatch(1);
     try {
       RunnableTaskRequest runnableTaskRequest =
         GSON.fromJson(request.content().toString(StandardCharsets.UTF_8), RunnableTaskRequest.class);
       className = runnableTaskRequest.getClassName();
       byte[] response = runnableTaskLauncher.launchRunnableTask(runnableTaskRequest);
-      responder.sendByteArray(HttpResponseStatus.OK, response, EmptyHttpHeaders.INSTANCE);
+
+      InputStream responseStream = new ByteArrayInputStream(response);
+
+      responder.sendContent(HttpResponseStatus.OK, new BodyProducer() {
+        @Override
+        public ByteBuf nextChunk() throws Exception {
+          byte[] buf = new byte[64 * 1024];
+          int len = responseStream.read(buf);
+          if (len == -1) {
+            return Unpooled.EMPTY_BUFFER;
+          }
+          return Unpooled.wrappedBuffer(buf, 0, len);
+        }
+
+        @Override
+        public void finished() {
+          Closeables.closeQuietly(responseStream);
+          latch.countDown();
+        }
+
+        @Override
+        public void handleError(@Nullable Throwable cause) {
+          LOG.error("Error when sending chunks", cause);
+        }
+      }, new DefaultHttpHeaders().add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM));
     } catch (ClassNotFoundException | ClassCastException ex) {
       responder.sendString(HttpResponseStatus.BAD_REQUEST, exceptionToJson(ex), EmptyHttpHeaders.INSTANCE);
+      latch.countDown();
     } catch (Exception ex) {
       LOG.error(String.format("Failed to run task %s",
-                              request.content().toString(StandardCharsets.UTF_8), ex));
+                              request.content().toString(StandardCharsets.UTF_8)), ex);
       responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex), EmptyHttpHeaders.INSTANCE);
+      latch.countDown();
     } finally {
+      try {
+        // wait until sending response is finished or an exception is caught.
+        if (!latch.await(SEND_RESPONSE_WAIT_TIME, TimeUnit.SECONDS)) {
```

Comment:
What's the purpose of this wait? You should use the `finished` and/or the `handleError` method for this purpose. Holding up an executor thread while streaming the content back is not good.